src/CattyTests/Download/StoreProjectsDownloaderTests.swift

```diff
@@ -575,8 +575,73 @@ class StoreProjectsDownloaderTests: XCTestCase {
         expect(downloader.fetchProjectDetails(for: project) { _, _ in }).toEventually(postNotifications(contain(expectedNotification)))
     }
 
-    private func getStoreProjectMock() -> StoreProject {
-        return StoreProject(projectId: "821",
+    // MARK: - Download project
+    func testDownloadProjectSucceeds() {
+           let dvrSession = Session(cassetteName: "StoreProjectDownloader.downloadProject.success")
+           let downloader = StoreProjectDownloader(session: dvrSession)
+           let project = getStoreProjectMock(projectId: "00fbfbe1-6f87-11ea-959c-000c293d4b76")
+           let expectation = XCTestExpectation(description: "Download Project")
+
+        downloader.download(project: project, completion: { data, error in
+            XCTAssertNil(error, "request failed")
+            guard data != nil else { XCTFail("no data received"); return }
+            expectation.fulfill()
+        }, progression: nil)
+
+            wait(for: [expectation], timeout: 1)
+    }
+
+    func testDownloadProjectFailsWithUnexpectedError() {
+        let mockSession = URLSessionMock()
+        let downloader = StoreProjectDownloader(session: mockSession)
+        let expectation = XCTestExpectation(description: "Download Projects")
+        let project = getStoreProjectMock()
+
+        downloader.download(project: project, completion: {_, error in
+            guard let error = error else { XCTFail("no error returned"); return }
+            XCTAssertEqual(error, .unexpectedError)
+            expectation.fulfill()
+        }, progression: nil)
+
+        wait(for: [expectation], timeout: 1)
+    }
+
+    func testDownloadProjectFailsWithRequestError() {
+        let dvrSession = Session(cassetteName: "StoreProjectDownloader.downloadProject.fail.request")
+        let downloader = StoreProjectDownloader(session: dvrSession)
+        let expectation = XCTestExpectation(description: "Download Projects")
+        let project = getStoreProjectMock(projectId: "00fbfbe1-6f87-11ea-959c-000c293d4b76")
+
+        downloader.download(project: project, completion: { _, error in
+            guard let error = error else { XCTFail("no error received"); return }
+            switch error {
+            case let .request(error: _, statusCode: statusCode):
+                XCTAssertEqual(statusCode, 404)
+            default:
+                XCTFail("wrong error received")
+            }
+            expectation.fulfill()
+        }, progression: nil)
+        wait(for: [expectation], timeout: 1)
+    }
+
+    func testDownloadProjectVaildProgression() {
+        let config = URLSessionConfiguration.ephemeral
+        let session = URLSession(configuration: config, delegate: nil, delegateQueue: nil)
```

Comment:
Please do not make real API requests. This would require a working internet connection for every test run. Also, our Jenkins build would fail whenever the API is being maintained or is unavailable for other reasons. Also, it is not necessary to make those requests in order to test your functionality here. Inject test doubles to your system under test and stub the methods responsible for the _dataTask_.

You can use `URLSessionMock` for that which is using a `URLSessionDataTaskMock`. In order to stub KVO and `countOfBytesExpectedToReceive` (`countOfBytesReceived`) it is necessary to extend those test doubles (e.g. with an optional _bytesReceived_ and _bytesTotal_ property which gets returned by the stubbed original methods, i.e. `countOfBytesExpectedToReceive` and `countOfBytesReceived`). Please let me know if you know what I mean and if you need any help.

Reply:
@m-herold  I am trying to do this inside the URLSessionDataTaskMock class 

`private let bytesReceived: Int64 = 600`
        `private let bytesTotal: Int64 = 1200`

        override var countOfBytesReceived: Int64 {
            return bytesReceived
        }

        override var countOfBytesExpectedToReceive: Int64 {
            return bytesTotal
        }
I have a couple of doubts 
1. Is my approach correct?
2. What I am unable to get is how to stub KVO.

Reply:
Very good approach. But instead of hard-coding those values, I would suggest to pass it (optionally) to the `URLSessionMock` and `URLSessionDataTaskMock` initializers. Then the test will also be even more readable giving specific documentation on how your progress is being calculated (not only > 0 and < 1 but also the specific calculation).

You can do KVO stubbing by telling the compiler that `countOfBytesReceived` is affected by your `bytesReceived`:
`@objc dynamic class func keyPathsForValuesAffectingCountOfBytesReceived() -> Set<String> { [ "bytesReceived" ] }`

You might have to add `@objc dynamic` to your properties (and maybe also `dynamic` to your overridden computed properties) to make KVO work.

Reply:
@m-herold I have updated progression observer option since we never change the value of `countOfBytesReceived` and `countOfBytesExpectedToReceive` beside setting it initially. Not sure if it is the correct approach. Please review.

Reply:
Your approach is good, although it implies that anyone using the `URLSessionMock` and _URLSessionDataTaskMock_ would have to use the option _initial_ for KVO which is not perfect. However, this seems to be sufficient for testing your PR and since this approach is a very simple solution, your changes here look good to me.